deepmd/infer/deep_pot.py

```diff
@@ -231,30 +262,20 @@ def eval(
             The atomic virial. Only returned when atomic == True
         """
         # reshape coords before getting shape
-        natoms = len(atom_types)
-        coords = np.reshape(np.array(coords), [-1, natoms * 3])
-        numb_test = coords.shape[0]
-        if atomic:
-            if self.modifier_type is not None:
-                raise RuntimeError('modifier does not support atomic modification')
-            if self.auto_batch_size is not None:
-                return self.auto_batch_size.execute_all(self._eval_inner, numb_test, natoms,
-                           coords, cells, atom_types, fparam = fparam, aparam = aparam, atomic = atomic, efield = efield)
-            return self._eval_inner(coords, cells, atom_types, fparam = fparam, aparam = aparam, atomic = atomic, efield = efield)
-        else :
-            if self.auto_batch_size is not None:
-                e, f, v = self.auto_batch_size.execute_all(self._eval_inner, numb_test, natoms,
-                              coords, cells, atom_types, fparam = fparam, aparam = aparam, atomic = atomic, efield = efield)
-            else:
-                e, f, v = self._eval_inner(coords, cells, atom_types, fparam = fparam, aparam = aparam, atomic = atomic, efield = efield)
-            if self.modifier_type is not None:
-                me, mf, mv = self.dm.eval(coords, cells, atom_types)
-                e += me.reshape(e.shape)
-                f += mf.reshape(f.shape)
-                v += mv.reshape(v.shape)
-            return e, f, v
+        natoms, numb_test = self._get_natoms_and_nframes(coords, atom_types)
+        output = self._eval_func(self._eval_inner, numb_test, natoms)(coords, cells, atom_types, fparam = fparam, aparam = aparam, atomic = atomic, efield = efield)
 
-    def _eval_inner(
+        if self.modifier_type is not None:
+            if atomic:
+                raise RuntimeError('modifier does not support atomic modification')
+            me, mf, mv = self.dm.eval(coords, cells, atom_types)
+            e, f, v = output[:3]
+            e += me.reshape(e.shape)
+            f += mf.reshape(f.shape)
+            v += mv.reshape(v.shape)
```

Comment:
`output` is not updated with me, mf, mv before return.

Reply:
I don't think so. Here is a minimal example:
```py
>>> import numpy as np
>>> output = (np.ones(3), np.ones(3), np.ones(3), np.ones(3), np.ones(3))
>>> e,f,v = output[:3]
>>> e += 1
>>> f += 2
>>> v += 3
>>> output
(array([2., 2., 2.]), array([3., 3., 3.]), array([4., 4., 4.]), array([1., 1., 1.]), array([1., 1., 1.]))
```

See https://stackoverflow.com/a/35910888/9567349. In Python, `a+=b` is different from `a=a+b`. `a=a+b` creates a new object, but `a+=b` modifies the original object.

Reply:
by ` e,f,v = output[:3]` e, f, v are copies of the list, not refs of the list element. You may do a test as 
```python
In [1]: a = [1,2,3]                                                             

In [2]: a[2]+=1                                                                 

In [3]: a                                                                       
Out[3]: [1, 2, 4]

In [4]: e,f,v=a[:3]                                                             

In [5]: e+=1                                                                    

In [6]: f+=1                                                                    

In [7]: v+=1                                                                    

In [8]: a                                                                       
Out[8]: [1, 2, 4]

In [9]: e, f, v                                                                 
Out[9]: (2, 3, 5)

```

Reply:
I can reproduce your example. 
It seems that python does not copy the numpy arrays , but do copy numbers. 
Adding modification in this way confuses the readers of the code, I suggest you changing it to 
```python
output[0] += me.reshape(e.shape)
...
```